runtime/compiler/trj9/optimizer/J9TransformUtil.cpp

```diff
@@ -1135,7 +1147,16 @@ bool
 J9::TransformUtil::transformIndirectLoadChainAt(TR::Compilation *comp, TR::Node *node, TR::Node *baseExpression, uintptrj_t *baseReferenceLocation, TR::Node **removedNode)
    {
    TR::VMAccessCriticalSection transformIndirectLoadChainAt(comp->fej9());
-   bool result = TR::TransformUtil::transformIndirectLoadChainImpl(comp, node, baseExpression, (void*)*baseReferenceLocation, removedNode);
+   void *baseAddress;
+   if (baseExpression->getOpCode().hasSymbolReference() && baseExpression->getSymbol()->isStatic())
+      {
+      baseAddress = (void *)comp->fej9()->getStaticReferenceFieldAtAddress(comp, (uintptrj_t)baseReferenceLocation, baseExpression->getSymbolReference());
+      }
+   else
+      {
+      baseAddress = (void *)comp->fej9()->getReferenceAtAddress((uintptrj_t)baseReferenceLocation);
+      }
+   bool result = TR::TransformUtil::transformIndirectLoadChainImpl(comp, node, baseExpression, baseAddress, removedNode);
```

Comment:
Depending on the caller, `*baseReferenceLocation` may be the address of e.g. a `J9Class` rather than an object. Is it safe to run the read barrier in that case?

Reply:
I could be wrong, but because this is an address that does not point to the heap (and not a 64-bit integer who's value might be an address into the heap), performing a read barrier shouldn't cause any problems. However, it's probably best confirmed by @amicic .